File: src/sens.c
/*
* Copyright (c) 2020 Vittorio Folino
*
* This file is part of lsensors.
*
* lsensors is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* lsensors is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.

* You should have received a copy of the GNU General Public License
* along with lsensors.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "sens.h"

list *getSensorsNames(sensList *sensors) {
    list *sensorsList = initList();
    if (sensorsList  == NULL)
        return NULL;
    char chipName[MAX_LEN];
    while (sensors != NULL) {
        if (sensors_snprintf_chip_name(chipName, MAX_LEN, sensors->name) < 0) {
            return NULL;
        } else if (addNode(sensorsList, chipName)) {
            return NULL;
        }
        sensors = sensors->next;
    }
    return sensorsList;
}

subfeaturesList 
*getAllSubfeatures(const sensors_chip_name *name, 
                const sensors_feature *feature) {

    int nr = 0;
    subfeaturesList *outList;
    subfeaturesList *start = NULL;
    const sensors_subfeature *subfeature = sensors_get_all_subfeatures(name, feature, &nr);
    if(subfeature != NULL) {
        outList = (subfeaturesList *) malloc(sizeof(subfeaturesList));
            if(outList == NULL)
                return NULL;
        start = outList;
    }
    while(subfeature != NULL) {
        outList->subfeature = subfeature;
        subfeature = sensors_get_all_subfeatures(name, feature, &nr);
        if(subfeature != NULL) {
            outList->next = (subfeaturesList *) malloc(sizeof(subfeaturesList));
            outList = outList->next;
            if(outList == NULL)
                return NULL;
        }
    }
    outList->next = NULL;
    return start;
}

featuresList *getAllFeatures(const sensors_chip_name *name) { 
    int nr = 0;
    featuresList *outList;
    featuresList *start = NULL;
    const sensors_feature *feature = sensors_get_features(name, &nr);
    if(feature != NULL) {
        outList = (featuresList *) malloc(sizeof(featuresList));
        if(outList == NULL)
            return NULL;
        start = outList;
    }
    while(feature != NULL) {
        outList->feature = feature;
        outList->subfeatures = getAllSubfeatures(name, feature);
        feature = sensors_get_features(name, &nr);
        if (feature != NULL) {
            outList->next = (featuresList *) malloc(sizeof(featuresList));
            outList = outList->next;
            if(outList == NULL)
                return NULL;
        }
    }
    outList->next = NULL;
    return start;
}

sensList *getAllSensors() {
    int nr = 0;
    sensList *sensorsList;
    sensList *start = NULL;
    const sensors_chip_name *name = sensors_get_detected_chips(NULL, &nr);
    if(name != NULL) {
        sensorsList = (sensList *) malloc(sizeof(sensList));
        if(sensorsList == NULL)
            return NULL;
        start = sensorsList;
    }
    while(name != NULL) {
        sensorsList->name = name;
        sensorsList->features = getAllFeatures(name);
        name = sensors_get_detected_chips(NULL, &nr);
        if(name != NULL) {
            sensorsList->next = (sensList *) malloc(sizeof(sensList));
            sensorsList = sensorsList->next;
            if(sensorsList == NULL) {
                return NULL;
            }
        }
    }
    sensorsList->next = NULL;
    return start;
}

int hasFeatureTemp(featuresList *features) {
    while(features != NULL) {
        if(features->feature->type == SENSORS_FEATURE_TEMP) {
            return 1;
        }
        features = features->next;
    }
    return 0;
}

void printSensorsValues(sensList *sensors, list *sensNames) {
    featuresList *features;
    subfeaturesList *subfeatures;
    int subfeat_nr;
    double value;
    wprintf(L"{");
	while(sensors != NULL) {
		wprintf(L"\n\t\"%s\": {", sensNames->string);
        sensNames = sensNames->next;
        features = sensors->features;
        
        while(features != NULL) {
            char *featureName = sensors_get_label(sensors->name, features->feature);
            wprintf(L"\n\t\t\"%s\": {", featureName);
            subfeatures = features->subfeatures;
            
            while(subfeatures != NULL) {
                subfeat_nr = subfeatures->subfeature->number;
                sensors_get_value(sensors->name, subfeat_nr, &value);
                wprintf(L"\n\t\t\t\"%s\": %lf", subfeatures->subfeature->name, value);
                if (subfeatures->next != NULL) {
                    wprintf(L",");
                }
                subfeatures = subfeatures->next;
            }
            wprintf(L"\n\t\t}");
            if (features->next != NULL) {
                wprintf(L",");
            }
            features = features->next;
            free(featureName);
        }
        wprintf(L"\n\t}");
        if(sensors->next != NULL) {
            wprintf(L",");
        }
        sensors = sensors->next;
	}
    wprintf(L"\n}\n");
}

void freeSubfeaturesList(subfeaturesList *list) {
    subfeaturesList *next;
    while(list != NULL) {
        next = list->next;
        free(list);
        list = next;
    }
}

void freeFeaturesList(featuresList *list) {
    featuresList *next;
    while(list != NULL) {
        next = list->next;
        freeSubfeaturesList(list ->subfeatures);
        free(list);
        list = next;
    }
}

void freeSensList(sensList *list) {
    sensList *next;
    while(list != NULL) {
        next = list->next;
        freeFeaturesList(list->features);
        free(list);
        list = next;
    }
}

File: lib/sens.py
#    Copyright (c) 2020 Vittorio Folino
#
#    This file is part of lsensors.
# 
#    lsensors is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    lsensors is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with lsensors.  If not, see <http://www.gnu.org/licenses/>.

import os
import sys
import json
import time
from datetime import datetime
import curses

BOUNDING = '-------------------------------------------------------------------------------'

def getTime():
    now = datetime.now()
    timestamp = int(datetime.timestamp(now))
    return datetime.fromtimestamp(timestamp)

class Sensors:

    def __init__(self, wd, settings: dict):
        sensBin = os.path.join('bin', 'sens')
        self.sens = os.path.join(wd, sensBin)

        self.avgValues = dict()
        self.minValues = dict()
        self.maxValues = dict()
        self.settings = settings
        self.samplingTime = settings["samplingTime"]
        if len(self.settings["sensorsList"]) == 0:
            self.updateSensors()
        else:
            self.sensorsList = settings["sensorsList"]
            self.sensors = settings["sensors"]

    def __updateSensorsList(self):
        outSens = os.popen(self.sens + ' list')
        self.sensorsList = outSens.readlines()
        self.sensorsList = [i.strip() for i in self.sensorsList]
        outSens.close()

    def rawOutput(self):
        outSens = os.popen(self.sens)
        sample = outSens.read()
        outSens.close()
        return sample
    
    def updateSensors(self):
        sensors = json.loads(self.rawOutput())
        self.__updateSensorsList()
        for sens in self.sensorsList:
            features = list(sensors[sens].keys())
            for feature in features:
                subfeatures = list(sensors[sens][feature])
                for subfeature in subfeatures:
                    sensors[sens][feature][subfeature] = False
        self.sensors = sensors
        self.settings["sensorsList"] = self.sensorsList
        self.settings["sensors"] = sensors
        return self.settings

    def __genValue(self, sensor, feature, subfeature):
        while True:
            out = json.loads(self.rawOutput())
            yield out[sensor][feature][subfeature]

    def getGens(self):
        gens = dict()
        for sensor in self.sensors.keys():
            for feature in self.sensors[sensor].keys():
                for subfeat in self.sensors[sensor][feature].keys():
                    if self.sensors[sensor][feature][subfeat]:
                        key = sensor + ' ' + feature + ' ' + subfeat
                        gens[key] = self.__genValue(sensor, feature, subfeat)
        return gens
    
    def print(self, sensor, feature, subfeature):
        while True:
            values = self.__genValue(sensor, feature, subfeature)
            print(next(values))
            time.sleep(self.samplingTime)
    
    def printAll(self):
        n_values = 0
        gens = self.getGens()
        screen = curses.initscr()
        curses.curs_set(0)
    
        try:
            while True:
                row = 0
                n_values += 1
                for sensor in gens.keys():
                    
                    value = next(gens[sensor])
                    if sensor not in self.minValues.keys():
                        self.minValues[sensor] = value
                        self.avgValues[sensor] = value
                        self.maxValues[sensor] = value
                    else:
                        self.minValues[sensor] = min(value, self.minValues[sensor])
                        self.avgValues[sensor] += value
                        self.maxValues[sensor] = max(value, self.maxValues[sensor])
                    screen.addstr(row, 0, sensor)
                    screen.addstr(row + 1, 0, 'value: ' + str(round(value, 1)) + ' °C')
                    screen.addstr(row + 1, 20, 'min: ' + str(round(self.minValues[sensor], 1)) + ' °C')
                    screen.addstr(row + 1, 40, 'avg: ' + str(round(self.avgValues[sensor]/n_values, 1)) + ' °C')
                    screen.addstr(row + 1, 60, 'max: ' + str(round(self.maxValues[sensor], 1)) + ' °C')
                    screen.refresh()
                    row += 3
                curses.napms(self.samplingTime * 1000)

        except (KeyboardInterrupt, SystemExit):
            curses.endwin()
    
    def logAll(self):
        n_values = 0
        logFolder = self.settings["logFolder"]
        filePath = os.path.join(logFolder, 'thermals.log')
        gens = self.getGens()
        fp = open(filePath, 'a')
        try:
            while True:
                fp.write(BOUNDING + '\n')
                fp.write(str(getTime()) + '\n')
                n_values += 1
                for sensor in gens.keys():
                    value = next(gens[sensor])
                    if sensor not in self.minValues.keys():
                        self.minValues[sensor] = value
                        self.avgValues[sensor] = value
                        self.maxValues[sensor] = value
                    else:
                        self.minValues[sensor] = min(value, self.minValues[sensor])
                        self.avgValues[sensor] += value
                        self.maxValues[sensor] = max(value, self.maxValues[sensor])

                    fp.write(sensor + '\n')
                    fp.write('value: ' + str(round(value, 1)) + ' °C\t')
                    fp.write('min: ' + str(round(self.minValues[sensor], 1)) + ' °C\t')
                    fp.write('avg: ' + str(round(self.avgValues[sensor]/n_values, 1)) + ' °C\t')
                    fp.write('max: ' + str(round(self.maxValues[sensor], 1)) + ' °C\n\n')
                fp.write(BOUNDING + '\n')
                time.sleep(self.samplingTime)
    
        except (KeyboardInterrupt, SystemExit):
            fp.close()
